packages/babel-parser/src/plugins/typescript/scope.ts

```diff
@@ -35,11 +38,50 @@ class TypeScriptScope extends Scope {
 // explanation of how typescript handles scope.
 
 export default class TypeScriptScopeHandler extends ScopeHandler<TypeScriptScope> {
+  importsStack: Set<string>[] = [];
+
   createScope(flags: ScopeFlags): TypeScriptScope {
+    this.importsStack.push(new Set()); // SCOPE_TS_TOP_LEVEL should be kept
+
     return new TypeScriptScope(flags);
   }
 
+  enter(flags: number): void {
+    if (flags == SCOPE_TS_MODULE) {
+      this.importsStack.push(new Set());
+    }
+
+    super.enter(flags);
+  }
+
+  exit() {
+    const flags = super.exit();
+
+    if (flags == SCOPE_TS_MODULE) {
+      this.importsStack.pop();
+    }
+  }
+
+  hasImport(name: string, allowShadow?: boolean) {
+    const len = this.importsStack.length;
+    if (this.importsStack[len - 1].has(name)) {
+      return true;
+    }
+    return !allowShadow && len > 1 && this.importsStack[0].has(name);
```

Comment:
The module declarations can be nested, so it may not suffice to check only the top level imports.

```ts
declare module 'test/sub1' {
  import type { JSONSchema7 } from 'json-schema';
  module 'test/sub2' {
    export { JSONSchema7 }
  }
}
```

Reply:
Good catch!